code/android-core-library/src/main/java/com/adobe/marketing/mobile/utils/RemoteDownloader.kt

```diff
@@ -0,0 +1,130 @@
+/*
+  Copyright 2022 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+ */
+
+package com.adobe.marketing.mobile.utils
+
+import com.adobe.marketing.mobile.LoggingMode
+import com.adobe.marketing.mobile.MobileCore
+import com.adobe.marketing.mobile.internal.utility.StringUtils
+import com.adobe.marketing.mobile.services.CacheFileService
+import com.adobe.marketing.mobile.services.Networking
+import java.io.File
+
+/**
+ * Facilitates downloading content from a URL into a local cache directory.
+ * Stores downloaded content in to a directory governed by [CacheFileService].
+ *
+ * Currently, triggers to [download] are independent, successive calls.
+ * This class is intentionally final and components are required to use this via composition;
+ * This allows easier path for components using this class to add queuing logic for jobs internally,
+ * maintaining compatibility when this class adds queuing logic in the future, and facilitating
+ * cleaner tests in composing components.
+ *
+ * TODO: Refactor this class to either eliminate default caching or make caching optional.
+ */
+class RemoteDownloader(
+    private val networkService: Networking,
+    private val cacheFileService: CacheFileService
+) {
+
+    companion object {
+        private const val TAG = "RemoteDownloader"
+    }
+
+    enum class Reason {
+        INVALID_URL,
+        RESPONSE_PROCESSING_FAILED,
+        CANNOT_WRITE_TO_CACHE_DIR,
+        NOT_MODIFIED,
+        NO_DATA,
+        SUCCESS
+    }
+
+    data class RemoteDownloadResult(val data: File?, val reason: Reason)
+
+    /**
+     * Represents a component that provides metadata about file content.
+     */
+    interface MetadataProvider {
+        companion object MetadataKeys {
```

Comment:
IMO, I don't think we will go to the fully kotlin sdk in the near future. At least, part of the existing internal extensions will still be java extensions for a long time. And I'm not optimistic about the above reason No 2 as Kotlin is a superset of Java which, I guess,  may be difficult to debug the java classes implementing kotlin interfaces. If we can expose public interfaces in Java, I think it could save us some debug time in the future.

Reply:
Done. Based on our slack discussion, I made the public facing interfaces to be Java only 👍